pkg/search/postgres/store.go

```diff
@@ -131,13 +136,97 @@ func NewGenericStoreWithPermissionChecker[T any, PT unmarshaler[T]](
 			return setPostgresOperationDurationTime
 		}(),
 		permissionChecker: checker,
+		cache:             nil,
+		useCache:          false,
 	}
 }
 
+// NewGenericStoreWithCache returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStoreWithCache[T any, PT unmarshaler[T]](
+	db postgres.DB,
+	schema *walker.Schema,
+	pkGetter primaryKeyGetter[T, PT],
+	insertInto inserter[T, PT],
+	copyFromObj copier[T, PT],
+	setAcquireDBConnDuration durationTimeSetter,
+	setPostgresOperationDurationTime durationTimeSetter,
+	upsertAllowed upsertChecker[T, PT],
+	targetResource permissions.ResourceMetadata,
+) *GenericStore[T, PT] {
+	store := &GenericStore[T, PT]{
+		db:          db,
+		schema:      schema,
+		pkGetter:    pkGetter,
+		insertInto:  insertInto,
+		copyFromObj: copyFromObj,
+		setAcquireDBConnDuration: func() durationTimeSetter {
+			if setAcquireDBConnDuration == nil {
+				return doNothingDurationTimeSetter
+			}
+			return setAcquireDBConnDuration
+		}(),
+		setPostgresOperationDurationTime: func() durationTimeSetter {
+			if setPostgresOperationDurationTime == nil {
+				return doNothingDurationTimeSetter
+			}
+			return setPostgresOperationDurationTime
+		}(),
+		upsertAllowed:  upsertAllowed,
+		targetResource: targetResource,
+		cache:          make(map[string]PT),
+		useCache:       true,
+	}
+	store.resetCache(sac.WithAllAccess(context.Background()))
+	return store
+}
+
+// NewGenericStoreWithPermissionCheckerWithCache returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStoreWithPermissionCheckerWithCache[T any, PT unmarshaler[T]](
+	db postgres.DB,
+	schema *walker.Schema,
+	pkGetter primaryKeyGetter[T, PT],
+	insertInto inserter[T, PT],
+	copyFromObj copier[T, PT],
+	setAcquireDBConnDuration durationTimeSetter,
+	setPostgresOperationDurationTime durationTimeSetter,
+	checker walker.PermissionChecker,
+) *GenericStore[T, PT] {
+	store := &GenericStore[T, PT]{
+		db:          db,
+		schema:      schema,
+		pkGetter:    pkGetter,
+		copyFromObj: copyFromObj,
+		insertInto:  insertInto,
+		setAcquireDBConnDuration: func() durationTimeSetter {
+			if setAcquireDBConnDuration == nil {
+				return doNothingDurationTimeSetter
+			}
+			return setAcquireDBConnDuration
+		}(),
+		setPostgresOperationDurationTime: func() durationTimeSetter {
+			if setPostgresOperationDurationTime == nil {
+				return doNothingDurationTimeSetter
+			}
+			return setPostgresOperationDurationTime
+		}(),
+		permissionChecker: checker,
+		cache:             make(map[string]PT),
+		useCache:          true,
+	}
+	store.resetCache(sac.WithAllAccess(context.Background()))
+	return store
+}
+
 // Exists tells whether the ID exists in the store.
 func (s *GenericStore[T, PT]) Exists(ctx context.Context, id string) (bool, error) {
 	defer s.setPostgresOperationDurationTime(time.Now(), ops.Exists)
```

Comment:
Feel like there should be special metrics _explicitly_ for the cache part of a store when caching is enabled. This would be another argument for making the cache a more standalone interface / abstraction instead of trying to put it too close to the generic store.